core/src/main/scala/akka/persistence/r2dbc/journal/R2dbcJournal.scala

```diff
@@ -147,27 +157,62 @@ private[r2dbc] final class R2dbcJournal(config: Config, cfgPath: String) extends
 
       serialized match {
         case Success(writes) =>
-          journalDao.writeEvents(writes).map(_ => Success(()))(ExecutionContexts.parasitic)
-        case Failure(t) =>
-          Future.successful(Failure(t))
+          journalDao.writeEvents(writes)
+        case Failure(exc) =>
+          Future.failed(exc)
       }
     }
 
     val persistenceId = messages.head.persistenceId
-    val writeResult =
+    val writeResult: Future[Option[Instant]] =
       if (messages.size == 1)
-        atomicWrite(messages.head).map(_ => Nil)(ExecutionContexts.parasitic)
+        atomicWrite(messages.head)
       else {
         // persistAsync case
         // easiest to just group all into a single AtomicWrite
         val batch = AtomicWrite(messages.flatMap(_.payload))
-        atomicWrite(batch).map(_ => Nil)(ExecutionContexts.parasitic)
+        atomicWrite(batch)
+      }
+
+    val writeAndPublishResult: Future[Done] =
+      publish(messages, writeResult)
```

Comment:
Doesn't this mean we could publish even if write fails? Shouldn't it be onComplete for writeResult rather?

Reply:
inside `def publish` it is `flatMap` on the `writeResult`

Reply:
I can rename that `publish` to `publishAfterWrite`

Reply:
Ah, the flatMap inside `publish` is what I was missing. I think it's ok like this.